API_MANIFESTO.md

```diff
@@ -0,0 +1,91 @@
+# The API Manifesto
+
+# Infinispan API dependency module
+
+The ```infinispan-api``` dependency will contain the public API. This module will be independent from the other modules. 
+Dependency with implementations will be **logical**. This will keep the modularity needed for the different use cases.
+
+Examples:
+ * Cache use case
+ 
+ * NoSQL data store use case
+ 
+ * Tasks, concurrency, counters uses cases
+
+The dependencies to thirty-party libraries must be avoided. This module should contain dependencies to JDK classes, 
+not more.
+
+# User First
+
+* Before making an API public, consider to test the usability with someone not familiar with the feature
+
+* Convention over Configuration: Make sure a default behaviour exists. Don't make me think.
+
+* Avoid exposing internals as much as possible. Even if your feature lays on top of a cache today, internals can change.
+
+* Each public API/Feature must have a simple tutorial showcasing the example. Without a demo, the API does not exist
+
+* If you are unsure about an API, mark it as ```@Experimental```. 
+
+* Don't let ```@Experimental``` tag last forever. 
+
+* Ask for user's feedback actively (zulip, user forums, blog post, twitter, github ...)
+
+* Be clear about whats PUBLIC API and what internals/implementations are. 
+
+
+# Client/Server First
+
+* New interfaces should be the same for embedded and client/server
+
+* Always start defining your API considering users in client/server mode. Embedded implementation will follow.
+
+* If a particular need is demanded for embedded, before doing something specific
+   - Discuss with the team the need. Is it really mandatory?
+   - Is this a general purpose or a really advanced need? 
+   - Will the functionality be widely used? Why is this embedded only then?
+   
+Embedded only API should be an exception at this point and they might not live in the **infinispan-api** module.
+
+
+# When do we create a new module/dependency?
+
+This has to be discussed case by case with the team. Several factors need to be considered.
+
+* Can the implementation of a new API live in an already existing module?
+
+* Would be better for the user if we create a new module, but we add the needed dependencies for the user?
+
+
+# Creating API
+
+* When you create a new API, you should consider the use case it's about to solve. 
+
+* Avoid Swiss knifes on interfaces. Keep it focuses and small enough.
+
+* Remember public APIs are like diamonds, they last forever. Rarely we will be allowed to just remove an API without 
+breaking compatibility. 
+
+
+# Evolving API
+
+* Adding a new method to an existing interface must be done with care. Consider the need of the method versus creating a
+new interface.
+
+* If there is no other choice that adding a new method, provide a **default implementation**
+
+* Methods tend to need parameter overloading overtime. Use a **Parameters** object to assure evolution over time instead 
```

Comment:
Actually, I think overloading can lead to explosion of methods, even with one overloaded extra method per each method. I'm more keen of something like `with(...).<my_method>()` approach, where a single method can be added for N methods. It all depends on the use case. Maybe you can provide some examples here?

Also, you should consider sometimes the granularity of the overload urges. For example, in our case, could we have avoided expiration/lifespan method overloads if the granularity for defining expiration/lifespan would exclusively have remained at the cache level?

Reply:
I think I don't totally get de `with` approach right now. Could you show me an example API existing, please?

In our case, for expiration and lifespan, if the granularity was cache level only, yes, we could have avoided all. With the params object approach, we could have avoided some of them but not all.

Reply:
@karesti Sorry for missing your reply, comments:

> I think I don't totally get de with approach right now. Could you show me an existing API example, please?

Our `FunctionalMap.withParams()` implements this pattern. It serves both as a replacement of `Cache.withFlag()` which was limited to only enums, and also as a replacement of the overloads that take lifespan/maxIdle. 

Indeed, by having cache level granularity for lifespan/maxIdle we would have avoided the overloading, but there's plenty of things where it can useful: e.g. pass in version to apply to a given write, e.g. `withParams(new VersionParam(x)).put(...)`, or to give specific lock timeout to a call `withParams(new LockTimeoutParam(y)).put(...)`.

I'm not necessarily advocating for `withParams` name but something along those lines. I find it's a neat way to pass in cross-cutting optional parameters.